packages/ckeditor5-engine/tests/model/utils/insertcontent.js

```diff
@@ -1299,6 +1299,1251 @@ describe( 'DataController utils', () => {
+					setData( model, '[]<paragraph>foo</paragraph>' );
+
+					insertHelper( '<imageBlock></imageBlock>', {
+						'marker-a': { start: [ 0 ], end: [ 0 ] }
+					} );
+
+					const expectedMarker = model.markers.get( 'marker-a' );
+
+					// <paragraph></paragraph>[<imageBlock></imageBlock>]<paragraph>foo</paragraph>
+					expect( getData( model ) ).to.equal( '<paragraph></paragraph>[<imageBlock></imageBlock>]<paragraph>foo</paragraph>' );
+
+					// <paragraph>{}</paragraph><imageBlock></imageBlock><paragraph>foo</paragraph>
+					expect( expectedMarker.getRange().start.path ).to.deep.equal( [ 0, 0 ] );
+					expect( expectedMarker.getRange().end.path ).to.deep.equal( [ 0, 0 ] );
+				} );
+
+				it( 'should create multiple collapsed marker after inserting different elements in empty content', () => {
+					setData( model, '' );
+
+					insertHelper( '<imageInline></imageInline><imageBlock></imageBlock>', {
+						'marker-a': { start: [ 0 ], end: [ 0 ] },
+						'marker-b': { start: [ 1 ], end: [ 1 ] }
+					} );
+
+					const expectedMarkerA = model.markers.get( 'marker-a' );
+					const expectedMarkerB = model.markers.get( 'marker-b' );
+
+					// <paragraph><imageInline></imageInline></paragraph>[<imageBlock></imageBlock>]
+					expect( getData( model ) ).to.equal( '<paragraph><imageInline></imageInline></paragraph>[<imageBlock></imageBlock>]' );
+
+					// <paragraph>{}<imageInline></imageInline>()</paragraph><imageBlock></imageBlock>
+					expect( expectedMarkerA.getRange().start.path ).to.deep.equal( [ 0, 0 ] );
+					expect( expectedMarkerA.getRange().end.path ).to.deep.equal( [ 0, 0 ] );
+					expect( expectedMarkerB.getRange().start.path ).to.deep.equal( [ 0, 1 ] );
+					expect( expectedMarkerB.getRange().end.path ).to.deep.equal( [ 0, 1 ] );
+				} );
+			} );
+
+			it( 'should does not create marker when insertion is forbidden', () => {
+				setData( model, '<wrapper><limit><paragraph>[]</paragraph></limit></wrapper>', {
+					'marker-a': { start: [ 0 ], end: [ 1 ] }
+				} );
+
+				const expectedMarker = model.markers.get( 'marker-a' );
+
+				// Pasted content is forbidden in current selection.
+				insertHelper( '<wrapper><limit><paragraph>foo</paragraph></limit></wrapper>' );
+
+				// <wrapper><limit><paragraph>[]</paragraph></limit></wrapper>
+				expect( getData( model ) ).to.equal( '<wrapper><limit><paragraph>[]</paragraph></limit></wrapper>' );
+
+				expect( expectedMarker ).to.be.null;
+			} );
+
+			it( 'should create 500 markers', () => {
+				setData( model, '<paragraph>foo[]</paragraph>' );
+
+				const mockMarkers = {};
+				for ( let i = 0; i < 500; i++ ) {
+					mockMarkers[ `comment-thread-${ i }` ] = { start: [ 0, 0 ], end: [ 0, 3 ] };
+				}
+
+				insertHelper( '<paragraph>Bar</paragraph>', mockMarkers );
+
+				for ( let i = 0; i < 500; i++ ) {
+					expect( model.markers.get( `comment-thread-${ i }` ) ).to.exist;
+				}
+			} );
+
+			describe( 'affected range', () => {
```

Comment:
I thought you forgot about testing `affectedRange` :). I thought you will add these checks in current tests.

Reply:
Can we stay with these tests?

Reply:
My problem is that these test cover much less scenarios than the rest of the test you added. And it will be easier to add `affectedRange` checks to the tests you wrote above, than to recreated those scenarios here.

Our current solution looks like it should not change `affectedRange`. I would be fine with adding tests for `affectedRange` in a followup issue. In this case, we can leave this tests suite, add an issue and when solving that new issue, we will remove this suite and make `affectedRange` checks in other markers-related tests.

BTW. It may be easier for you to check `affectedRange` if you use `stringify()` (from the same file as `getData()`). You can pass there a range that should be printed. It would be also easier for checking markers. I should have pointed this out earlier :(.